src/services/ProveGatewayService.ts

```diff
@@ -0,0 +1,181 @@
+import BigNumber from 'bignumber.js';
+import { GatewayRepository } from '../repositories/GatewayRepository';
+import { MessageRepository } from '../repositories/MessageRepository';
+import Logger from '../Logger';
+import Utils from '../Utils';
+import { AUXILIARY_GAS_PRICE } from '../Constants';
+import Observer from '../observer/Observer';
+import { AuxiliaryChain } from '../repositories/AuxiliaryChainRepository';
+
+const Mosaic = require('@openst/mosaic.js');
+
+export default class ProveGatewayService extends Observer<AuxiliaryChain> {
+  private gatewayRepository: GatewayRepository;
+
+  private messageRepository: MessageRepository;
+
+  private originWeb3: object;
+
+  private auxiliaryWeb3: object;
+
+  private auxiliaryWorkerAddress: string;
+
+  private gatewayAddress: string;
+
+  private auxiliaryChainId: number;
+
+  /**
+   *  Constructor
+   *
+   * @param gatewayRepository Instance of auxiliary chain repository.
+   * @param messageRepository Instance of message repository.
+   * @param originWeb3 Origin Web3 instance.
+   * @param auxiliaryWeb3 Auxiliary Web3 instance.
+   * @param auxiliaryWorkerAddress auxiliary worker address, this should be
+   *                               unlocked and added in web3 wallet.
+   * @param gatewayAddress Address of gateway contract on origin chain.
+   * @param auxiliaryChainId Auxiliary chain Id.
+   */
+  public constructor(
+    gatewayRepository: GatewayRepository,
+    messageRepository: MessageRepository,
+    originWeb3: object,
+    auxiliaryWeb3: object,
+    auxiliaryWorkerAddress: string,
+    gatewayAddress: string,
+    auxiliaryChainId: number,
+  ) {
+    super();
+
+    this.gatewayRepository = gatewayRepository;
+    this.originWeb3 = originWeb3;
+    this.auxiliaryWeb3 = auxiliaryWeb3;
+    this.auxiliaryWorkerAddress = auxiliaryWorkerAddress;
+    this.gatewayAddress = gatewayAddress;
+    this.messageRepository = messageRepository;
+    this.auxiliaryChainId = auxiliaryChainId;
+
+    // Suppressing lint error
+    // Refer: https://github.com/typescript-eslint/typescript-eslint/issues/636
+    /* eslint-disable @typescript-eslint/unbound-method */
+    this.reactTo = this.reactTo.bind(this);
+    this.update = this.update.bind(this);
+  }
+
+
+  /**
+   * This method react on changes in auxiliary chain models.
+   * @param auxiliaryChains List of auxiliary chains model
+   */
+  public async update(auxiliaryChains: AuxiliaryChain[]): Promise<void> {
+    const proveGatewayPromises = auxiliaryChains
```

Comment:
Should we instead just get the single auxiliary chain that equals to stored auxiliaryChainId (we can assert that there is not more than one). After that implementation below is simplified as we do not need to filter and map. I believe, Promise.all is not needed also in this case.

Reply:
It's inforced by the `update` method from the interface. It will always receive list because in future it can react to multiple anchor handler.

Reply:
My point is the following. Yes, update() might give a list of auxiliary chains. However, the current service is only interested on a single auxiliary chain (that's what we do in a filter part).

My point was, as we currently, only taking an auxiliary chain with chain id equal to this.auxiliaryChainId, is that easy to get that single auxiliary chain and assert that it's maximum one?

Reply:
The current implementation is more robust. It might be tricky for the repository to provide only one record to service. Here we can easily filter out.

Reply:
Done